microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/config/java/MicroProfileConfigPropertyDiagnosticsParticipant.java

```diff
@@ -0,0 +1,137 @@
+import org.eclipse.lsp4mp.jdt.core.utils.AnnotationUtils;
+import org.eclipse.lsp4mp.jdt.internal.metrics.MicroProfileMetricsConstants;
+
+/**
+ * Validate "defaultValue" attribute of @ConfigProperty and generate
+ * diagnostics if "defaultValue" cannot be represented by the given
+ * field type.
+ *
+ * See https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/converters.asciidoc
+ * for more details on default converters.
+ */
+public class MicroProfileConfigPropertyDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
+
+	@Override
+	public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProgressMonitor monitor)
+			throws CoreException {
+		List<Diagnostic> diagnostics = new ArrayList<>();
+		validateClass(diagnostics, context, monitor);
+		return diagnostics;
+	}
+
+	private void validateClass(List<Diagnostic> diagnostics, JavaDiagnosticsContext context,
+			IProgressMonitor monitor) {
+		CompilationUnit ast = context.getASTRoot();
+		ast.accept(new ConfigPropertiesAnnotationValidator(diagnostics, context));
+	}
+
+	public class ConfigPropertiesAnnotationValidator extends ASTVisitor {
+		private List<Diagnostic> diagnostics;
+		private JavaDiagnosticsContext context;
+
+		public ConfigPropertiesAnnotationValidator(List<Diagnostic> diagnostics, JavaDiagnosticsContext context) {
+			this.diagnostics = diagnostics;
+			this.context = context;
+		}
+
+		@Override
+		public boolean visit(NormalAnnotation node) {
+			if (AnnotationUtils.isMatchAnnotation(node, MicroProfileConfigConstants.CONFIG_PROPERTY_ANNOTATION)) {
+				try {
+					Expression defValueExpr = AnnotationUtils.getAnnotationMemberValueExpression(node, MicroProfileConfigConstants.CONFIG_PROPERTY_ANNOTATION_DEFAULT_VALUE);
+					ASTNode parent = node.getParent();
+					if (defValueExpr != null && parent instanceof FieldDeclaration) {
+						String defValue = ((StringLiteral) defValueExpr).getLiteralValue();
+						FieldDeclaration field = (FieldDeclaration) parent;
+						ITypeBinding fieldBinding = field.getType().resolveBinding();
+						if (fieldBinding != null && !isAssignable(fieldBinding, defValue)) {
+							Range range = context.getUtils().toRange(context.getTypeRoot(), defValueExpr.getStartPosition(), defValueExpr.getLength());
+							diagnostics.add(new Diagnostic(range,
+									"'" + defValue + "'" + " does not match the expected type of '" + fieldBinding.getName() + "'.",
+									DiagnosticSeverity.Error, MicroProfileMetricsConstants.DIAGNOSTIC_SOURCE));
+						}
+					}
+				} catch (JavaModelException e) {
+					return false;
+				}
+			}
+			return false;
+		}
+
+		private boolean isAssignable(ITypeBinding fieldBinding, String defValue) {
+			String fqn = Signature.getTypeErasure(fieldBinding.getQualifiedName());
+			try {
+				switch (fqn) {
+				case "boolean":
+				case "java.lang.Boolean":
+					return Boolean.valueOf(defValue) != null;
```

Comment:
There is a problem with boolean value (you can set any value, and you will have not error).

It's because `Boolean.valueOf("XXXXX") `return `false` and not null. Please check boolean like https://github.com/eclipse/lsp4mp/blob/master/microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/services/properties/PropertiesFileValidator.java#L291

Reply:
Ah, could it be that `PropertiesFileValidator.isBooleanString(..)` is too strict ? According to https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/converters.asciidoc :

`boolean and java.lang.Boolean , values for true (case insensitive) "true", "1", "YES", "Y" "ON". Any other value will be interpreted as false`

So even "XXXXX" is technically just false. If that's the case, then any String value should be supported.

Reply:
I think we should follow the specification, so you are right and PropertiesFileValidator.isBooleanString should be less strict. Thanks @rgrunber for your clarification.